jena-shacl/src/main/java/org/apache/jena/shacl/lib/ShLib.java

```diff
@@ -0,0 +1,209 @@
+                , "       sh:resultMessage ?message ;"
+                , "       sh:resultSeverity  ?severity ; "
+                , "       ."
+                , "    OPTIONAL { ?R sh:sourceConstraintComponent ?component }"
+                , "    OPTIONAL { ?R sh:sourceShape ?sourceShape }"
+                , "    OPTIONAL { ?R sh:resultPath    ?path}"
+                ,"}");
+        try ( QueryExecution qExec = QueryExecutionFactory.create(qs, report.getModel()) ) {
+            ResultSet rs = qExec.execSelect();
+            if ( ! rs.hasNext() ) {
+                System.out.println("No violations");
+            } else {
+                rs.forEachRemaining(row->{
+                    // XXX Rebuild report entry
+//                    ReportEntry e = ReportEntry.create()
+//                        .focusNode(focusNode)
+//                        .resultPath(path)
+//                        .message(item.getMessage())
+//                        .severity(Severity.Violation)
+//                        .source(shape.getShapeNode())
+//                        .sourceConstraint(constraint)
+//                        //.detail(null)
+//                        .value(focusNode)
+//                        .triple(triple)
+//                        ;
+
+                    RDFNode focusNode = row.get("focusNode");
+                    String msg = row.getLiteral("message").getLexicalForm();
+                    RDFNode path = row.get("path");
+                    if ( path != null )
+                        System.out.printf("Node=%s   Path=%s\n", displayStr(focusNode), path);
+                    else
+                        System.out.printf("Node=%s\n",displayStr(focusNode));
+                    System.out.printf("  %s\n", msg);
+                });
+            }
+        }
+    }
+
+    static String displayStr(RDFNode n) {
+        return displayStr(n.asNode(), nodeFmt);
+    }
+
+    public static String displayStr(Node n) {
+        return displayStr(n, nodeFmt);
+    }
+
+    public static String displayStr(Node n, NodeFormatter fmt) {
+        IndentedLineBuffer sw = new IndentedLineBuffer() ;
+        nodeFmt.format(sw, n);
+        return sw.toString() ;
+    }
+
+    public static boolean isImmediate(Target target) {
+        TargetType targetType = target.getTargetType();
+        return targetType.equals(TargetType.targetObjectsOf) || targetType.equals(TargetType.targetSubjectsOf);
+    }
+
+    public static Node focusNode(Triple triple, Target target) {
+        switch(target.getTargetType()) {
+            //case targetClass :
+            case targetNode :
+                return target.getObject();
+            case targetObjectsOf :
+                return triple.getObject();
+            case targetSubjectsOf :
+                return triple.getSubject();
+            default :
+        }
+        return null;
+    }
+
+    @SafeVarargs
+    public static <X> X getFirstNonNull(X...xs) {
```

Comment:
Maybe we could use `ObjectUtils.firstNonNull` from Commons Lang (used in jena-arq I think?)